src/main/java/com/aws/greengrass/FleetProvisioningByClaimPlugin.java

```diff
@@ -284,54 +293,104 @@
         }
     }
 
-    private void copyPrivateKey(String rootPath, String privateKeyPath) {
-        Path permPrivateKeyPath = Paths.get(rootPath, PRIVATE_KEY_PATH_RELATIVE_TO_ROOT);
-        try {
-            if (Files.notExists(permPrivateKeyPath)) {
-                Files.createFile(permPrivateKeyPath);
+    private static Path getRootPath(Map<String, Object> parameterMap) {
+        return getPath(parameterMap, ROOT_PATH_PARAMETER_NAME);
+    }
+
+    private static Path getPath(Map<String, Object> parameterMap, String key) {
+        if (parameterMap.get(key) == null) {
+            throw new DeviceProvisioningRuntimeException(
+                    String.format("Config parameter %s is missing", key));
+        }
+        return getPath(parameterMap.get(key).toString());
+    }
+
+    private static Path getPathOrDefaultFromRoot(Map<String, Object> parameterMap, String key, String defaultName) {
+        if (parameterMap.get(key) == null) {
+            // if not provided, result path relative to rootPath
+            Path rootPath = getRootPath(parameterMap);
+            try {
+                return  rootPath.resolve(defaultName);
+            } catch (InvalidPathException e) {
+                throw new DeviceProvisioningRuntimeException(
+                        String.format("Invalid %s: %s. rootPath: %s",
+                                key, defaultName, rootPath));
             }
-            Files.copy(Paths.get(privateKeyPath), permPrivateKeyPath, StandardCopyOption.REPLACE_EXISTING);
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true)
-                    .ownerWrite(true).build(), permPrivateKeyPath);
-        } catch (IOException e) {
-            logger.atError().log("Caught exception while copying private key to root directory");
-            throw new DeviceProvisioningRuntimeException("Failed to copy private key to root directory", e);
+        } else {
+            return getPath(parameterMap.get(key).toString());
         }
     }
 
-    private void writeCertificateAndKeyToPath(CreateKeysAndCertificateResponse response, String rootPath) {
+    private static Path getPath(String path) {
         try {
-            Path certPath = Paths.get(rootPath, DEVICE_CERTIFICATE_PATH_RELATIVE_TO_ROOT);
-            if (Files.notExists(certPath)) {
-                Files.createFile(certPath);
-            }
-            Files.write(certPath, response.certificatePem.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true)
-                .ownerWrite(true).build(), certPath);
+            return Paths.get(path).normalize();
+        } catch (InvalidPathException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Consider passing the exception object either as  &#34;rootCause&#34; or as an inner exception parameter, to the constructor of the new exception before throwing new exception. This will help keep stack trace.
```
@@ -328,3 +328,3 @@
             throw new DeviceProvisioningRuntimeException(
-                    String.format(&#34;Invalid path: %s&#34;, path));
+                    String.format(&#34;Invalid path: %s&#34;, path), e);
         }

```